src/Data/Array/Core.idr

```diff
@@ -44,12 +44,21 @@ at (IA ad) m =
 ||| a new size index.
 |||
 ||| Note: If you only need a small portion of a potentially large
-|||       array the resto of which you no longer need, consider to
+|||       array the rest of which you no longer need, consider to
 |||       release the large array from memory by invoking `force`.
 export
 take : (0 m : Nat) -> IArray n a -> {auto 0 lte : LTE m n} -> IArray m a
 take _ (IA arr) = IA arr
 
+||| We can drop n elements from the start of an array. O(n)
+|||
+||| Note: If you only need a small portion of a potentially large
+|||       array the rest of which you no longer need, consider to
+|||       release the large array from memory by invoking `force`.
+export
+drop : (0 m : Nat) -> IArray n a -> {auto 0 lte : LTE m n} -> IArray (minus n m) a
+drop _ (IA arr) = IA arr
```

Comment:
This seems wrong. It is just a copy of the implementation of `take`. This should actually be the implementation that is used by the version of `take` that operates on `Array`.

The following should work here:

```idris
export
drop : {n : _} -> (m : Nat) -> IArray n a -> IArray (n `minus` m) a
drop m arr = generate (n `minus` m) (\f => at arr (inc f))
```

where `inc` has type

```idris
export
inc : {m : _} -> Fin (n `minus` m) -> Fin n
```

Now, the challenge will be that `inc` must be O(1) in order for `drop` to be O(n). Therefore, the implementation of `inc` must go via `finToNat` and back to `Nat` via `natToFinLT` after adding `m`. This will require some proof handling.

Reply:
I can help you with this, if you get stuck.

Reply:
I think this makes sense to me, although I think if we put this `drop` in `Data.Array.Core` then there is a cyclical import issue when trying to use `generate` here.

Reply:
@stefan-hoeck

I have this so far:

```
||| We can drop n elements from the start of an array. O(n)
|||
||| Note: If you only need a small portion of a potentially large
|||       array the rest of which you no longer need, consider to
|||       release the large array from memory by invoking `force`.
export
drop : {n : _} -> (m : Nat) -> IArray n a -> IArray (n `minus` m) a
drop m (IA arr) = generate (n `minus` m) (\f => at arr (inc f))
  where
    inc : {m : _} -> Fin (n `minus` m) -> Fin n
    inc o with ((plus m 1) <= n) proof eq
      _ | True  = inc o $ natToFinLT (plus m 1) @{lteOpReflectsLTE _ _ eq}
      _ | False = natToFinLT m @{lteOpReflectsLTE _ _ eq}
```

But I think I'm getting confused with the `inc` function implementation a bit, could you help me understand this?

Reply:
You won't need a `with` expression here: We *know* that `finToNat n' + m` is strictly less than `n`. Here's how I did it:

In `Data.Array.Index`:

```idris

export
0 minusLT : (x,m,n : Nat) -> LT x (n `minus` m) -> LT (x+m) n
minusLT x _     0     y = absurd y
minusLT x 0     (S k) y = rewrite plusZeroRightNeutral x in y
minusLT x (S k) (S j) y =
  let p1 := minusLT x k j y
   in LTESucc (rewrite sym (plusSuccRightSucc x k) in p1)

export
inc : {m : _} -> Fin (n `minus` m) -> Fin n
inc x =
  let 0 p1 := finToNatLT x
   in natToFinLT (finToNat x + m) @{minusLT _ _ _ p1}
```

And in `Data.Array.Indexed`:

```idris
export
drop : {n : _} -> (m : Nat) -> IArray n a -> IArray (n `minus` m) a
drop m arr = generate (n `minus` m) (\f => at arr (inc f))
```

Reply:
Ah I see, this makes sense (move the `drop` implementation to `Data.Array.Indexed`, and then call this implementation for the `drop` in `Data.Array`.

Seeing your `minusLT` implementation definitely helps make sense of the proofs needed.